### ISSUE TITLE
Non-deterministic sort misaligns rolling/Elo features onto the wrong fixture (41% of training rows, ~5pp accuracy loss)

### ISSUE THREAD

**Issue by abivan100-stack** [Raghav2012Code/epl-predictor#3]
## Summary
Rolling-form and Elo features are merged onto the **wrong fixture** whenever matches share a date, which is every real gameweek. Measured on the repo's own cached data, **41% of training rows carry another fixture's features**.

## Root cause
`build_engineered_dataset` establishes a canonical order and a positional key:

```python
raw_matches = raw_matches.copy().sort_values(by="date").reset_index(drop=True)   # feature_engineering.py:469
raw_matches["match_id"] = raw_matches.index                                      # :470
```

Two helpers then **re-sort by `date` and re-derive positions from scratch**, ignoring that `match_id`:

* `transform_matches_to_team_perspective` — `df.copy().sort_values(by="date").reset_index(drop=True)` (`:300`) then `m_ids = df.index.values` (`:302`)
* `compute_dynamic_elo` — `matches_df.sort_values(by="date").reset_index(drop=True)` (`:141`), returning plain lists in *its* order, which the caller assigns **positionally** at `:491-495`

`DataFrame.sort_values` defaults to `kind="quicksort"`, which is **not stable** and **not idempotent on already-sorted input with ties**. So the merge at `:484-487` (`on="match_id"`) joins the form of match `O2[k]` onto match `O1[k]` — a different fixture played the same day.

```
rows permuted by an idempotent re-sort: 1026 / 2280   (data/raw, 2020-2026)
```

## Measured impact
Against an independent ground-truth recomputation over the real 2280-match dataset:

| feature | wrong rows |
|---|---|
| `home_roll_points_5` | 936 / 2264 (41.3%) |
| `home_rest_days` | 587 / 2264 (25.9%) |
| `home_elo` | 1026 / 2280 (45.0%) |

Isolation: after giving every match a unique timestamp (`date + i minutes`, preserving date order) and re-running the **unmodified** code, `home_roll_points_5 wrong: 0 / 2264`.

Model cost (XGBoost, split 2024-01-01, everything else identical):

| | val accuracy | val log-loss |
|---|---|---|
| as-is (tied dates) | 0.4640 | 1.0531 |
| de-tied (order-correct) | **0.5138** | **1.0115** |

**~5.0pp of accuracy is being lost to this bug**, and every committed artifact (`models/metrics.json`, `data/predictions_2026_2027.*`, `visuals/*`) is produced from the corrupted matrix.

## Concrete example
2024-08-17 — Ipswich v Liverpool, Arsenal v Wolves and Everton v Brighton share a date. Arsenal's row can be trained with Everton's rolling goals/points/rest-days and Ipswich's pre-match Elo.

Note this is **not** future leakage (same date, no lookahead) — it is a wholesale scrambling of the feature matrix.

## Suggested fix
1. Sort with a deterministic total order everywhere: `sort_values(by=["date", "home_team", "away_team"], kind="mergesort")`. `(date, home_team, away_team)` is unique per match, so the result is idempotent and every helper agrees.
2. In `transform_matches_to_team_perspective`, reuse the caller's `match_id` column when present instead of re-deriving it from `df.index`.
3. Have `compute_dynamic_elo` return values aligned to `match_id` rather than as positional lists.

---
*Found during a systematic audit of `dc1156b`. Every finding was verified by executing the code, not by reading alone; the commands and measured numbers are included above.*
